polkadot/runtime/parachains/src/mock.rs

```diff
@@ -388,6 +388,121 @@ impl ValidatorSetWithIdentification<AccountId> for MockValidatorSet {
 	type IdentificationOf = FoolIdentificationOf;
 }
 
+pub mod mock_assigner {
+	use super::*;
+	pub use pallet::*;
+
+	#[frame_support::pallet]
+	pub mod pallet {
+		use super::*;
+
+		#[pallet::pallet]
+		#[pallet::without_storage_info]
+		pub struct Pallet<T>(_);
+
+		#[pallet::config]
+		pub trait Config: frame_system::Config + configuration::Config + paras::Config {}
+
+		#[pallet::storage]
+		pub(super) type MockAssignerQueue<T: Config> =
+			StorageValue<_, VecDeque<V0Assignment>, ValueQuery>;
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn add_on_demand_order(
+			order: V0Assignment,
+			location: QueuePushDirection,
+		) -> Result<(), DispatchError> {
+			MockAssignerQueue::<T>::try_mutate(|queue| {
+				// Don't need to worry about max assignment queue size in scheduler tests
+				match location {
+					QueuePushDirection::Back => queue.push_back(order),
+					QueuePushDirection::Front => queue.push_front(order),
+				};
+				Ok(())
+			})
+		}
+
+		fn is_legacy_core(core_idx: &CoreIndex) -> bool {
+			let parachain_cores = <paras::Pallet<Test>>::parachains().len() as u32;
+			(0..parachain_cores).contains(&core_idx.0)
+		}
+	}
+
+	impl<T: Config> AssignmentProvider<BlockNumber> for Pallet<T> {
+		type AssignmentType = V0Assignment;
+		type OldAssignmentType = V0Assignment;
+		// Format has not changed for parachains, therefore still version 0.
+		const ASSIGNMENT_STORAGE_VERSION: AssignmentVersion = AssignmentVersion::new(0);
+
+		fn migrate_old_to_current(
+			old: Self::OldAssignmentType,
+			_core: CoreIndex,
+		) -> Self::AssignmentType {
+			old
+		}
+
+		// This can matter, even in tests. Using condensed form of calculation from actual assigner.
+		fn session_core_count() -> u32 {
+			let config = <configuration::Pallet<Test>>::config();
+			(<paras::Pallet<Test>>::parachains().len() as u32)
+				.saturating_add(config.on_demand_cores)
+		}
+
+		// Create a new assignment to pass if core belongs to bulk buyer, else pop
+		// on demand assignment from front of queue.
+		fn pop_assignment_for_core(core_idx: CoreIndex) -> Option<Self::AssignmentType> {
+			if Self::is_legacy_core(&core_idx) {
```

Comment:
Why are we doing this?

Reply:
This assigner is fulfilling the role of either a `ParachainsAssigner` or an `OnDemandAssigner` for each call. By calling `is_legacy_core` we determine which way to act using the input `core_idx`. The idea is that for a bulk time core index this assigner will always behave like a `ParachainsAssigner`. The same applies to on-demand core indices.

The follow up question to this would be, "Why do we need the MockAssigner to act like any particular kind of assigner." The answer is that the assigner is how we control the flow of eventual `ParasEntries` in scheduler tests as well as the `AssignmentProviderConfig` telling scheduler tests how to handle them. 

We test the scheduler in scenarios where it needs to behave as if it has a real `OnDemandAssigner`: `on_demand_claims_are_pruned_after_timing_out` and `next_up_on_time_out_reuses_claim_if_nothing_queued`.

We also test it in scenarios where it needs to behave as if it has a `ParachainsAssigner`:
`next_up_on_time_out_is_parachain_always`

Reply:
In this specific instance, for on-demand core indices, we need to drain an entry from the queue each time so that we can test scenarios where the queue becomes empty.